src/tree/AATree.scala

```diff
@@ -0,0 +1,201 @@
+  }
+
+  def insert[B >: T: Ordering](a: B): AATree[B] = this match {
+    case Empty                                                 =>
+      Fork(a, 1, Empty, Empty)
+    case Fork(v, lv, l, r) if implicitly[Ordering[B]].lt(a, v) => 
+      Fork(v, lv, l.insert(a), r).rebalanceTree
+    case Fork(v, lv, l, r) if implicitly[Ordering[B]].gt(a, v) => 
+      Fork(v, lv, l, r.insert(a)).rebalanceTree
+    case _                                                     =>
+      this
+  }
+
+  def merge[B >: T: Ordering](that: AATree[B]): AATree[B] = 
+    if(this == Empty) that
+    else if(that == Empty) this
+    else {
+      left.merge(right.merge(that)).insert(value)
+    }
+
+
+  // Let's make our tree an instance of functor, and monad
+  def map[B: Ordering](f: T => B): AATree[B] = this match {
+    case Empty            => Empty
+    case Fork(v, _, l, r) =>
+      val mappedL = l.map(f)
+      val mappedR = r.map(f)
+      mappedL.merge(mappedR).insert(f(v))
+  }
+
+  def flatMap[B: Ordering](f: T => AATree[B]): AATree[B] = this match {
+    case Empty            => Empty
+    case Fork(v, _, l, r) =>
+      val mappedL = l.flatMap(f)
+      val mappedR = r.flatMap(f)
+      mappedL.merge(mappedR).merge(f(v))
+  }
+
+  def fold[B: Ordering](f: (T, B, B) => B, z: B): B = this match {
+    case Empty            => z
+    case Fork(v, _, l, r) =>
+      f(v, l.fold(f, z), r.fold(f, z))
+  }
+
+  def delete[B >: T: Ordering](a: B): AATree[B] = this match {
+    case Empty                                                 => 
+      Empty
+    case Fork(v, lv, l, r) if implicitly[Ordering[B]].gt(a, v) => 
+      Fork(v, lv, l, r.delete(a)).rebalanceAfterDelete
+    case Fork(v, lv, l, r) if implicitly[Ordering[B]].lt(a, v) => 
+      Fork(v, lv, l.delete(a), r).rebalanceAfterDelete
+    case Fork(_, _, Empty, Empty)                              =>
+      Empty
+    case Fork(_, lv, Empty, r)                                 =>
+      val succ = successor
+      val r2 = r.delete(succ)
+      Fork(succ, lv, Empty, r2).rebalanceAfterDelete
+    case Fork(_, lv, l, r)                                     =>
+      val pred = predecessor
+      val l2 = l.delete(pred)
+      Fork(pred, lv, l2, r).rebalanceAfterDelete
+  }
+
+  
+  // some helper functions
+  protected def rebalanceTree[B >: T: Ordering]: AATree[B] = skew.split
+
+  protected def rebalanceAfterDelete[B >: T: Ordering]: AATree[B] = {
+    val t1 = decreaseLevel.skew
+    val t2 = t1.right.skew
+    val t3 = t2 match {
+      case Fork(v1, lv1, l1, Fork(v2, lv2, l2, r2)) =>
+        Fork(v1, lv1, l1, Fork(v2, lv2, l2, r2.skew))
+      case _                                        =>
+        t2
+    }
+    val t4 = t3.split
+    t4 match {
+      case Fork(v, lv, l, r) => Fork(v, lv, l, r.split)
+      case _                 => t4
+    }
+  }
+
+  protected def successor[B >: T: Ordering]: B = {
+    def successor2(t: AATree[B]): B = this match {
```

Comment:
Yep. It's not visible outside `successor`, so we can use unified name for these functions. It usually `loop`.